### DISCUSSION TITLE
Accessing FieldInfo attributes from class in Pydantic V2

### DISCUSSION THREAD

**Discussion by DurandA** [pydantic/pydantic#6231] [Pydantic V2]
I want to create an ORM where models are defined as Pydantic models (similarly to [SQLModel](https://sqlmodel.tiangolo.com/), which uses Pydantic V1).

Here is an example of intended use:
```python
class Hero(MyExtendedPydanticModel):
    name: str = MyFieldInfo()
    team: Team|None = MyRelationshipFieldInfo()

h = await select(Hero).where(Hero.name == 'batman').load_related(Hero.team)
```
In order to support this API style—which is great for type hinting—it is required that attributes could be accessed both from the class and instances. For reference, [SQLModel overloads many Pydantic components](https://github.com/tiangolo/sqlmodel/blob/main/sqlmodel/main.py), including ModelMetaclass.

Is there a way to add access to attributes from both class and instances without overriding too much internals in Pydantic V2?

**Comment by adriangb** [comment]
@samuelcolvin @dmontagu we had discussed allowing this, it works with dataclasses but not BaseModel. I don't remember what the technical limitation we ran into was but I can dig into it a bit more if you agree we should make it work for BaseModel.


```python
from pydantic.dataclasses import dataclass


@dataclass
class Hero:
    name: str = "abc"

print(Hero.name == "xyz")  # False
```

**Comment by lig** [comment]
@DurandA There is `BaseModel.model_fields` class attribute that holds a dictionary of `FieldInfo` instances for each model field.

So, something like this could work:

```py
class ORMCOnditionBuilder:
    def __eq__(self) -> ORMCOnfition:
        ...

class ORMModelMeta(pydantic._internal._model_construction.ModelMetaclass):
    def __getattribute__(cls,  name):
        if name in cls.model_fields:
            return ORMCOnditionBuilder(cls, name, cls.model_fields[name])
        return super().__getattribute__(name)
        
class BaseORMModel(pydantic.BaseModel, metaclass=ORMModelMeta):
    pass
```

I think that this should work but there could be some cases of an unpredictable behavior with `ModelMetaclass` used this way. So, some additional logic could be required. It could be an option to try not to inherit from `pydantic.BaseModel` and use `pydantic_core` calls in your base class directly.

Also, I think that making `ModelMetaclass` a part of the `pydantic` public namespace could be a good idea. @samuelcolvin what do you think about making `ModelMetaclass` public?
